submissions/eoneoff/port-sniffer/sniffer.js

```diff
@@ -0,0 +1,198 @@
+                provided incorrectly the port range will be set to default full range (1-65535)
+
+args:Object                
+
+    module:bool     a bool argument, which tells the function to run as a module, that is to
+                    return a result as a list of open ports
+
+    silent:bool     a bool argument, which tells the function to supress the output and run
+                    in silent mode. Takes effect only if 'modlue' parameter is also set
+                    to 'true'
+
+The function is async and returns a promice. You shoud eather use it with await operator or
+with then() method`
+}
+
+const dns = require('dns').promises;
+const Socket = require('net').Socket;
+
+function* twister(){
+    let symbols = ['|','/','-','\\'];
+    let current = 0;
+    while (true) {
+        if (current == 4) current = 0;
+        yield symbols[current++];
+    }
+}
+
+function isIpAddress(address){
+    return address.search(/^((25[0-5]|2[0-4]\d|[01]?\d?\d)\.){3}(25[0-5]|2[0-4]\d|[01]?\d?\d)$/) == -1;
+}
+
+function isRange(ip) {
+    return ip.search(/\d{1,4}-\d{1,4}/) == -1;
+}
+
+function isInvalidPortsRange(portsRange){
+    return +portsRange[0] >= +portsRange[1] || +portsRange[0] < 1 || +portsRange[0] > 65535
+    || +portsRange[1] < 1 || +portsRange[1]>65535;
+}
+
+async function parseHost(hostName){
+    let host = '';
+    if(isIpAddress(hostName)){
+        try{
+            host = (await dns.lookup(hostName)).address;
+        } catch {
+            throw "Hostname can not be resolved";
+        }
+    } else {
+        host = hostName;
+    }
+    return host;
+}
+
+function parsePorts(ports){
+    if(isRange(ports)){
+        throw "Invalid port range argument"
+    } else {
+        let limits = ports.split("-");
+        if (isInvalidPortsRange(limits)){
+                throw "Port numbers out of range"
+            } else {
+                limits[1]++;
+                return limits;
+            }
+    }
+}
+
+async function parseArguments(args){
+    let host = "";
+    let ports = [1, 65536];
+    let hostIndex = args.indexOf('--host') + 1;
+    if(hostIndex == 0 || hostIndex == args.length){
+        throw "You must specify a host to scan ports";
+    } else {
+        host = await parseHost(args[hostIndex]);
+    }
+    let portsIndex = args.indexOf('--ports') + 1;
+    if (portsIndex != 0){
+        if(portsIndex == args.length){
+            console.log("Invalis ports specified, setting to default (1-65535)");
+        } else {
+            try{
+                ports = parsePorts(args[portsIndex]);
+            } catch {
+                console.log("Invalid ports specified, setting to default(1-65535)")
+            }
+        }
+    }
+
+    return {host, ports}
+}
+
+async function scan(host, ports, args = {silent:false,module:false}){
+    let openPorts = [];
+    let tw = twister();
+    process.stdout.write("scanning ");
+    for(let port = +ports[0]; port < ports[1]; ++port) {

```

Comment:
If you already using generators, I can advise more declarative style for your code.
Our task (besides validation), consist only from two steps:
1. Get the range of ports.
2. Filter each array of ports by async calls.
Consider this code:
```js
function* rangeInt(from, to) {
  for (; from <= to; from++) {
     yield from
  }
}

async function getOpenedPorts(from, to) {
   return Array.from(rangeInt(from, to)).filter(async (port) => isPortOpenned(port))
}
```
```js
// usage
const opennedPorts = await getOpenedPorts(1, 65536);
```
Note, you can also use spread with array literal instead of `Array.from`

Reply:
Oh, sorry, but I will argue that, Yes, it'll give a code a cool functional look, but we will trade if for memory and performance, because it'll actually be the same `for` cycle, but wrapped into generator.

Reply:
Performance/memory safe with your `for` loop is very poor, generally, irrelevant. But if you prefer `for` loops, keep it as is.